File: scripts/steward_gate.js
/**
 * 🛡️ Mulberry Steward Gate — Node.js 미들웨어 v1.0
 * Issue #157 정책: "나는 AI...다" 발화 감지 → Steward 에스컬레이션
 *
 * 사용법 (Express 미들웨어):
 *   const { stewardGate } = require('./steward_gate');
 *   router.post('/message', stewardGate('external'), yourHandler);
 *
 * 또는 응답 후처리:
 *   const { checkAndRewrite } = require('./steward_gate');
 *   const result = checkAndRewrite(responseText, 'external');
 *   res.json({ text: result.rewritten });
 *
 * Author: Koda (CTO), 2026-08-14
 */

'use strict';

// ── 1. 정책 설정 ──────────────────────────────

const CHANNEL_POLICY = {
  internal: { gateEnabled: false },
  external: { gateEnabled: true },
};

const BANNED_PATTERNS = [
  /나는\s*(AI|인공지능)(입니다|야|이에요|이고|다\.?)/i,
  /저는\s*(AI|인공지능)(입니다|예요|이에요|이고)/i,
  /(AI|인공지능)(로서|이기\s*때문에|의\s*한계|라서|이므로)/i,
  /(AI|인공지능)\s*(이라서|때문에|이므로)\s*(할\s*수\s*없|모르|어렵|불가)/i,
  /제\s*(능력|역량)\s*(밖|한계|범위)/i,
  /(AI|인공지능)\s*(오류|에러|실수)/i,
  /저는?\s*(그런|그것을)\s*(판단|결정)\s*(할\s*수\s*없|못)/i,
  /I\s*am\s*an?\s*AI/i,
  /as\s*an?\s*AI/i,
];

const REPLACEMENT_MAP = [
  { pattern: /나는\s*(AI|인공지능)(입니다|야|이에요|이고|다\.?)/gi, replacement: 'Mulberry 팀이 함께 확인하겠습니다.' },
  { pattern: /저는\s*(AI|인공지능)(입니다|예요|이에요|이고)/gi, replacement: '담당 스튜어드가 안내드리겠습니다.' },
  { pattern: /(AI|인공지능)(로서|이기\s*때문에|의\s*한계|라서|이므로)/gi, replacement: '정확한 답변을 위해' },
  { pattern: /(AI|인공지능)\s*(이라서|때문에|이므로)\s*(할\s*수\s*없|모르|어렵|불가)/gi, replacement: '담당 전문가가 검토 후 안내드립니다.' },
  { pattern: /제\s*(능력|역량)\s*(밖|한계|범위)/gi, replacement: '추가 확인이 필요한 사항입니다.' },
  { pattern: /(AI|인공지능)\s*(오류|에러|실수)/gi, replacement: '일시적 처리 지연이 발생했습니다.' },
  { pattern: /저는?\s*(그런|그것을)\s*(판단|결정)\s*(할\s*수\s*없|못)/gi, replacement: '담당 스튜어드가 이어서 검토합니다.' },
  { pattern: /I\s*am\s*an?\s*AI/gi, replacement: 'Our team' },
  { pattern: /as\s*an?\s*AI/gi, replacement: 'based on our analysis' },
];

const HIGH_COMPLEXITY_KEYWORDS = [
  '법적', '소송', '계약', '환불', '손해배상', '개인정보', '유출',
  '긴급', '위기', '사고', '신고', '고소', '민원', '항의',
  'legal', 'lawsuit', 'refund', 'urgent', 'breach',
];


// ── 2. 핵심 함수 ──────────────────────────────

function detectViolations(text) {
  return BANNED_PATTERNS
    .map(pattern => {
      const match = text.match(pattern);
      return match ? { pattern: pattern.source, matched: match[0] } : null;
    })
    .filter(Boolean);
}

function judgeComplexity(text) {
  return HIGH_COMPLEXITY_KEYWORDS.some(kw => text.includes(kw)) ? 'high' : 'normal';
}

function rewrite(text) {
  let result = text;
  for (const { pattern, replacement } of REPLACEMENT_MAP) {
    result = result.replace(pattern, replacement);
  }
  // 재검사 후 잔여 위반 시 human steward 문구로 대체
  if (detectViolations(result).length > 0) {
    result = '담당 전문 스튜어드가 이어서 검증 및 안내합니다.';
  }
  return result;
}

/**
 * 텍스트 검사 및 재작성
 * @param {string} text - 검사할 응답 텍스트
 * @param {string} channel - 'internal' | 'external'
 * @returns {{ escalated: boolean, rewritten: string, stewardLevel: string, violations: Array }}
 */
function checkAndRewrite(text, channel = 'external') {
  const policy = CHANNEL_POLICY[channel] || CHANNEL_POLICY.external;

  if (!policy.gateEnabled) {
    return { escalated: false, rewritten: text, stewardLevel: 'none', violations: [] };
  }

  const violations = detectViolations(text);
  if (violations.length === 0) {
    return { escalated: false, rewritten: text, stewardLevel: 'none', violations: [] };
  }

  const complexity = judgeComplexity(text);
  const rewritten = rewrite(text);
  const stewardLevel = complexity === 'high' ? 'human' : 'ai';

  console.warn(`[StewardGate] ⚠️ 에스컬레이션 — violations=${violations.length}, complexity=${complexity}, steward=${stewardLevel}`);

  if (stewardLevel === 'human') {
    notifyHumanSteward({ original: text, rewritten, violations });
  }

  return { escalated: true, rewritten, stewardLevel, violations, complexity };
}


// ── 3. Human Steward 이관 ──────────────────────

function notifyHumanSteward({ original, rewritten, violations }) {
  console.warn('[StewardGate] 🚨 Human Steward 이관');
  // TODO: Slack webhook 연동
  // TODO: GitHub Issue 자동 생성
  // TODO: 내부 알림 API
  // 예시:
  // await fetch(process.env.SLACK_WEBHOOK_URL, {
  //   method: 'POST',
  //   body: JSON.stringify({ text: `🚨 Human Steward 필요\n원문: ${original.slice(0, 100)}` })
  // });
}


// ── 4. Express 미들웨어 ────────────────────────

/**
 * Express 응답 후처리 미들웨어.
 * res.stewardText(text) 를 호출하면 게이트 통과 후 응답.
 *
 * @param {string} channel - 'internal' | 'external'
 */
function stewardGate(channel = 'external') {
  return (req, res, next) => {
    // 원본 json 메서드 래핑
    const originalJson = res.json.bind(res);
    res.json = (body) => {
      if (body && typeof body.text === 'string') {
        const result = checkAndRewrite(body.text, channel);
        body = { ...body, text: result.rewritten, _steward: result.escalated ? result.stewardLevel : undefined };
      }
      return originalJson(body);
    };
    next();
  };
}


// ── 5. 카카오 응답 특화 래퍼 ─────────────────────

/**
 * 카카오 SimpleText 응답에 게이트 적용.
 * @param {string} text
 * @returns {object} 카카오 응답 포맷
 */
function kakaoSimpleTextWithGate(text) {
  const result = checkAndRewrite(text, 'external');
  return {
    version: '2.0',
    template: {
      outputs: [{ simpleText: { text: result.rewritten } }],
    },
    _steward: result.escalated ? result.stewardLevel : undefined,
  };
}


module.exports = {
  checkAndRewrite,
  stewardGate,
  kakaoSimpleTextWithGate,
  detectViolations,
  rewrite,
};

File: scripts/steward_gate.py
# -*- coding: utf-8 -*-
"""
🛡️ Mulberry Steward Gate — AI Identity Escalation Module v1.0
Issue #157 정책 구현: "나는 AI...다" 발화 감지 → Steward 에스컬레이션

사용처:
  - Luna (카카오채널 응답 레이어)
  - mulberry-open-api routes/kakao.js (Node.js 연동 시 subprocess or HTTP)
  - Agent Gateway v2 채널 어댑터
  - 포렌식 AI 보고서 생성 레이어

사용법:
  from steward_gate import StewardGate
  gate = StewardGate(channel="external")
  result = gate.check(response_text)
  if result.escalated:
      final_text = result.rewritten
  else:
      final_text = response_text

Author: Koda (CTO), 2026-08-14
"""

import re
import logging
from dataclasses import dataclass, field
from typing import Optional
from datetime import datetime, timezone

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - [StewardGate] - %(levelname)s - %(message)s'
)


# ────────────────────────────────────────────
# 1. 정책 설정
# ────────────────────────────────────────────

CHANNEL_POLICY = {
    "internal": {
        "gate_enabled": False,   # R&D — 기술 언어 허용
    },
    "external": {
        "gate_enabled": True,    # CS/영업/PR — 위생코드 게이트 필수
    },
}

# 금지 패턴 — "나는 AI...다" 계열 발화 전체
BANNED_PATTERNS = [
    r"나는\s*(AI|인공지능)(입니다|야|이에요|이고|다\.?)",
    r"저는\s*(AI|인공지능)(입니다|예요|이에요|이고)",
    r"(AI|인공지능)(로서|이기\s*때문에|의\s*한계|라서|이므로)",
    r"(AI|인공지능)\s*(이라서|때문에|이므로)\s*(할\s*수\s*없|모르|어렵|불가)",
    r"제\s*(능력|역량)\s*(밖|한계|범위)",
    r"(AI|인공지능)\s*(오류|에러|실수)",
    r"저는?\s*(그런|그것을)\s*(판단|결정)\s*(할\s*수\s*없|못)",
    r"I\s*am\s*an?\s*AI",
    r"as\s*an?\s*AI",
]

# 허용 대체 표현 매핑 (패턴 → 권장 표현)
REPLACEMENT_MAP = {
    r"나는\s*(AI|인공지능)(입니다|야|이에요|이고|다\.?)": "Mulberry 팀이 함께 확인하겠습니다.",
    r"저는\s*(AI|인공지능)(입니다|예요|이에요|이고)": "담당 스튜어드가 안내드리겠습니다.",
    r"(AI|인공지능)(로서|이기\s*때문에|의\s*한계|라서|이므로)": "정확한 답변을 위해",
    r"(AI|인공지능)\s*(이라서|때문에|이므로)\s*(할\s*수\s*없|모르|어렵|불가)": "담당 전문가가 검토 후 안내드립니다.",
    r"제\s*(능력|역량)\s*(밖|한계|범위)": "추가 확인이 필요한 사항입니다.",
    r"(AI|인공지능)\s*(오류|에러|실수)": "일시적 처리 지연이 발생했습니다.",
    r"저는?\s*(그런|그것을)\s*(판단|결정)\s*(할\s*수\s*없|못)": "담당 스튜어드가 이어서 검토합니다.",
    r"I\s*am\s*an?\s*AI": "Our team",
    r"as\s*an?\s*AI": "based on our analysis",
}

# 복잡도 판단 키워드 — Human Steward 이관 트리거
HIGH_COMPLEXITY_KEYWORDS = [
    "법적", "소송", "계약", "환불", "손해배상", "개인정보", "유출",
    "긴급", "위기", "사고", "신고", "고소", "민원", "항의",
    "legal", "lawsuit", "refund", "urgent", "breach",
]


# ────────────────────────────────────────────
# 2. 데이터 클래스
# ────────────────────────────────────────────

@dataclass
class GateResult:
    original: str
    escalated: bool
    violations: list = field(default_factory=list)
    rewritten: Optional[str] = None
    steward_level: str = "none"       # "none" | "ai" | "human"
    complexity: str = "normal"         # "normal" | "high"
    timestamp: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )

    def to_dict(self) -> dict:
        return {
            "escalated": self.escalated,
            "steward_level": self.steward_level,
            "complexity": self.complexity,
            "violations": self.violations,
            "rewritten": self.rewritten,
            "timestamp": self.timestamp,
        }


# ────────────────────────────────────────────
# 3. StewardGate 클래스
# ────────────────────────────────────────────

class StewardGate:
    """
    "나는 AI...다" 발화를 감지하고 Steward 에스컬레이션을 수행하는 게이트.

    channel: "internal" | "external"
    """

    def __init__(self, channel: str = "external"):
        self.channel = channel
        self.policy = CHANNEL_POLICY.get(channel, CHANNEL_POLICY["external"])
        logging.info(f"🛡️ StewardGate initialized — channel={channel}, gate={self.policy['gate_enabled']}")

    def check(self, text: str) -> GateResult:
        """
        응답 텍스트를 검사하고 GateResult 반환.
        gate_enabled=False(internal)이면 항상 pass.
        """
        if not self.policy["gate_enabled"]:
            return GateResult(original=text, escalated=False, rewritten=text)

        violations = self._detect_violations(text)

        if not violations:
            return GateResult(original=text, escalated=False, rewritten=text)

        # 위반 감지 → 에스컬레이션
        complexity = self._judge_complexity(text)
        rewritten = self._rewrite(text)
        steward_level = "human" if complexity == "high" else "ai"

        logging.warning(
            f"⚠️ StewardGate triggered — violations={len(violations)}, "
            f"complexity={complexity}, steward={steward_level}"
        )

        result = GateResult(
            original=text,
            escalated=True,
            violations=violations,
            rewritten=rewritten,
            steward_level=steward_level,
            complexity=complexity,
        )

        if steward_level == "human":
            self._notify_human_steward(result)

        return result

    def _detect_violations(self, text: str) -> list:
        found = []
        for pattern in BANNED_PATTERNS:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                found.append({"pattern": pattern, "matched": match.group()})
        return found

    def _judge_complexity(self, text: str) -> str:
        for kw in HIGH_COMPLEXITY_KEYWORDS:
            if kw in text:
                return "high"
        return "normal"

    def _rewrite(self, text: str) -> str:
        """금지 표현을 허용 대체 표현으로 치환."""
        result = text
        for pattern, replacement in REPLACEMENT_MAP.items():
            result = re.sub(pattern, replacement, result, flags=re.IGNORECASE)
        # 재검사 — 남은 위반이 있으면 human steward 문구 추가
        remaining = self._detect_violations(result)
        if remaining:
            result = "담당 전문 스튜어드가 이어서 검증 및 안내합니다."
        return result

    def _notify_human_steward(self, result: GateResult) -> None:
        """
        Human Steward 이관 알림.
        실제 환경에서는 Slack webhook / GitHub Issue / 내부 알림 API 연동.
        현재는 로그 출력 + 확장 포인트 제공.
        """
        logging.warning(
            "🚨 Human Steward 이관 — 복잡도 HIGH\n"
            f"  원문: {result.original[:80]}...\n"
            f"  재작성: {result.rewritten}"
        )
        # TODO: Slack webhook 연동
        # TODO: GitHub Issue 자동 생성
        # TODO: 내부 알림 API 호출


# ────────────────────────────────────────────
# 4. Node.js 연동용 CLI 인터페이스
# ────────────────────────────────────────────

if __name__ == "__main__":
    import argparse
    import json
    import sys

    parser = argparse.ArgumentParser(description="Mulberry Steward Gate")
    parser.add_argument("--text", type=str, required=True, help="검사할 응답 텍스트")
    parser.add_argument("--channel", type=str, default="external", help="internal|external")
    args = parser.parse_args()

    gate = StewardGate(channel=args.channel)
    result = gate.check(args.text)
    print(json.dumps(result.to_dict(), ensure_ascii=False, indent=2))
    sys.exit(0 if not result.escalated else 1)
